File: src/python/presence_detector.py
from time import monotonic
from  enum import Enum
import RPi.GPIO as gpio
from led_strip import LedStrip


class State(Enum):
    IDLE = 0
    DETECTED = 1
    WAIT = 2


class Pir:
    DETECTED_PERIOD_S = 60
    WAIT_PERIOD_S = 5
    def __init__(self, pin: int, led_strip: LedStrip):
        self.detect_pin = pin  # BCM pin numbering
        self.start_time = 0
        self.state = State.IDLE
        self.led_strip = led_strip
        self.__init_gpio()

    def reset(self):
        self.state = State.IDLE
        self.led_strip.turn_off()

    def poll(self):
        if self.state == State.IDLE:
            if gpio.input(self.detect_pin) == gpio.HIGH:
                self.led_strip.turn_on()
                self.start_time = monotonic()
                self.state = State.DETECTED
            return
        if self.state == State.DETECTED:
            if gpio.input(self.detect_pin) == gpio.HIGH:
                self.start_time = monotonic()
                return
            if (monotonic() - self.start_time) >= self.DETECTED_PERIOD_S:
                self.led_strip.turn_off()
                self.start_time = monotonic()
                self.state = State.WAIT
            return
        if self.state == State.WAIT:
            if (monotonic() - self.start_time) >= self.WAIT_PERIOD_S:
                self.start_time = monotonic()
                self.state = State.IDLE

    def __init_gpio(self):
        gpio.setmode(gpio.BCM)
        gpio.setup(self.detect_pin, gpio.IN, pull_up_down=gpio.PUD_DOWN)

File: cpp/src/presence_detector.cpp
#include <wiringPi.h>
#include "presence_detector.hpp"

#define DETECTED_PERIOD_MS std::chrono::milliseconds(60000)
#define WAIT_PERIOD_MS std::chrono::milliseconds(5000)

void PresenceDetector::poll(void)
{
    switch (currentState)
    {
    case IDLE:
        if (isDetectPinAsserted()) {
            ledStrip_.setDutyCycle(100);
            timer_.setDelay(DETECTED_PERIOD_MS);
            currentState = DETECTED;
        }
        break;
    
    case DETECTED:
        if (isDetectPinAsserted()) {
            timer_.reset();
            return;
        }

        if (timer_.hasElapsed()) {
            ledStrip_.turnOff();
            timer_.setDelay(WAIT_PERIOD_MS);
            currentState = WAIT;
        }
        break;
    
    case WAIT:
        if (timer_.hasElapsed()) {
            currentState = IDLE;
        }
        break;
    
    default:
        break;
    }
}

void PresenceDetector::reset(void)
{
    ledStrip_.turnOff();
    currentState = IDLE;
}

bool PresenceDetector::isDetectPinAsserted(void)
{
    int value = digitalRead(detectPin_);

    return value == HIGH;
}